### PR TITLE
Refactor Object Detail Component into Typescript

### PR DESCRIPTION
**Pull request by iethree**
## Changes

- Object Detail Component is now split into sensible functional components
- Components are in typescript
- Components have much better unit test coverage
- No visual or functional changes

## Limitations

- I have not typed everything here. There are certain properties (like `settings`) where, because of the extensive use of prop spreading, it is nearly impossible to tell what all the possible properties or values are. I tried to strike a balance of making large improvements to the component structure, but not chasing down all the possible forms of data it could receive.
- I made an attempt at a storybook story, but couldn't get it to compile :cry: 
- I didn't move anything to emotion - in the interest of moving forward, I didn't rewrite existing css styles in emotion, also, I suspect much of these will change, and hope to write new styling in emotion

### CODE REVIEW

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetailsTable.tsx [hunk 1]
```diff
@@ -0,0 +1,138 @@
+import React from "react";
+import ExpandableString from "metabase/query_builder/components/ExpandableString";
+import { isID } from "metabase/lib/schema_metadata";
+import { TYPE, isa } from "metabase/lib/types";
+import { formatValue, formatColumn } from "metabase/lib/formatting";
+import cx from "classnames";
+import { t } from "ttag";
```
Comment (kulyk) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetailsTable.tsx:7]: Please sort imports according to our [style guide](https://www.metabase.com/docs/latest/developers-guide/frontend.html#javascript-conventions)

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectRelationships.tsx [hunk 2]
```diff
@@ -0,0 +1,108 @@
+import React from "react";
+import IconBorder from "metabase/components/IconBorder";
+import LoadingSpinner from "metabase/components/LoadingSpinner";
+import { foreignKeyCountsByOriginTable } from "metabase/lib/schema_metadata";
+import { inflect } from "inflection";
+import Icon from "metabase/components/Icon";
+import { t } from "ttag";
+import cx from "classnames";
+import { ForeignKey } from "metabase-types/types/ForeignKey";
```
Comment (kulyk) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectRelationships.tsx:9]: Please sort imports according to our [style guide](https://www.metabase.com/docs/latest/developers-guide/frontend.html#javascript-conventions)

**Review by kulyk — COMMENTED** [review]
Thanks for cleaning things up 👏 Left a few improvement ideas

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx [hunk 3]
```diff
@@ -0,0 +1,353 @@
...
+    // load up FK references
+    if (tableForeignKeys) {
+      loadObjectDetailFKReferences();
+    }
+    window.addEventListener("keydown", onKeyDown, true);
+
+    return () => {
+      window.removeEventListener("keydown", onKeyDown, true);
+    };
+  }, []); // eslint-disable-line react-hooks/exhaustive-deps
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:128]: I generally like smaller `useEffects` as well, but in the case of the hook that runs on mount, I think it aids readability to keep it all together - particularly here where we've got early return logic that we'd have to duplicate several times.

File: frontend/src/metabase-types/types/ForeignKey.ts [hunk 4]
```diff
@@ -0,0 +1,9 @@
+import { Field } from "./Field";
+
+export interface ForeignKey {
+  destination: Field;
+  destination_id: number;
+  origin: Field;
+  origin_id: number;
+  relationship: string; // enum?
+}
```
[the remark this answers is not in the document]
Comment (iethree) [reply frontend/src/metabase-types/types/ForeignKey.ts:9]: good to know!

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetailsTable.tsx (hunk 1 above)
Comment (iethree) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetailsTable.tsx:7]: thanks for pointing me to the style guide. very helpful!

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx (hunk 3 above)
[the remark this answers is not in the document]
Comment (kulyk) [reply frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:128]: Oh right, thanks for explaining that. Please take a look at the `useOnMount` though, it helps to get rid of the `eslint-disable-line react-hooks/exhaustive-deps` thing :) But that's nip for sure

File: frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx [hunk 5]
```diff
@@ -0,0 +1,351 @@
...
+import Icon from "metabase/components/Icon";
+import { NotFound } from "metabase/containers/ErrorPages";
+import { useOnMount } from "metabase/hooks/use-on-mount";
+import { usePrevious } from "metabase/hooks/use-previous";
+
+import { getObjectName, getIdValue } from "./utils";
+import { DetailsTable } from "./ObjectDetailsTable";
+import { Relationships } from "./ObjectRelationships";
+
+import Tables from "metabase/entities/tables";
```
Comment (kulyk) [inline frontend/src/metabase/visualizations/components/ObjectDetail/ObjectDetail.tsx:22]: Could you move the local imports to the end of the list, please?

**Review by kulyk — APPROVED** [review]
Sorry to bother you with the order of the imports, but I've just left a very last comment on that 🙈 Anyway, great work and thank you for cleaning things up 👏

**Pull request merged**